### PR TITLE
Patient Service Enhancements for Richer OpenEMR and FHIR R4 Responses

### PR DESCRIPTION
**Pull request by dixonwhitmire**
Fixes #
First iteration for #3353 

#### Short description of what this resolves:
This PR adds additional fields to the PatientService "get" functions to support a richer/more complete FHIR R4 Patient Resource response. I'll follow up on the related FHIR changes in a separate PR.

#### Changes proposed in this pull request:
This PR includes the following changes
- Additional columns in PatientService getOne and getAll to support richer API and FHIR R4 responses. Additional columns include:
    - county
    - drivers_license
    - contact_relationship (emergency contact name)
    - phone_home
    - phone_biz
    - phone_cell
    - status (marital status)
- Refactored PatientService.getAll to streamline SQL generation
- patient/:pid GET returns a 404 if resource is not found
- API and FHIR API README Updates pertaining to token requests

Additionally, I am happy to add test cases for these changes. If I could have guidance on whether I should "mock" the database records or install them as fixtures with a "real" database.

Finally, any PHP pointers or best practices are certainly appreciated.

Thanks!

### CODE REVIEW

**Comment by dixonwhitmire** [comment]
Hi @sjpadgett  - here's my first PR to pave the way for enriching the FHIR Patient Resource and auto-publishing.

File: src/RestControllers/PatientRestController.php [hunk 1]
```diff
@@ -19,6 +19,24 @@ class PatientRestController
...
+        "city",
+        "state",
+        "phone_home",
+        "phone_biz",
+        "phone_cell",
+        "email",
+        "DOB"
+    );
+
     public function __construct($pid)
```
Comment (dixonwhitmire) [inline src/RestControllers/PatientRestController.php:40]: I didn't want to tackle this in my first PR, but are there any thoughts on removing the dependency on the `$pid` in the constructor? We could just pass the $pid in as a parameter via the Rest Controller configuration. Just bringing this up as this seems to be the only controller which uses this technique/pattern.

Comment (sjpadgett) [reply src/RestControllers/PatientRestController.php:40]: Yep, it can be moved out of constructor to getOne($pid). Then service modified.

File: src/Services/PatientService.php [hunk 2]
```diff
@@ -268,27 +251,37 @@ public function getAll($search)
...
+                        phone_contact,
+                        phone_home,
+                        phone_biz,
+                        phone_cell,
+                        email,
+                        DOB,
+                        sex,
+                        race,
+                        ethnicity,
+                        status
```
Comment (sjpadgett) [inline src/Services/PatientService.php:284]: will need to get into PatientResource I assume

File: src/RestControllers/PatientRestController.php (hunk 1 above)
Comment (dixonwhitmire) [reply src/RestControllers/PatientRestController.php:40]: Thanks! I'll take a look at that change in the next iteration (shining up the patient FHIR endpoint)

File: src/Services/PatientService.php (hunk 2 above)
Comment (dixonwhitmire) [reply src/Services/PatientService.php:284]: Yes! That is correct. 

In the next PR I'm going to review the existing code to verify that we're supporting the "standard" [FHIR Search Parameters](https://github.com/openemr/openemr/pull/3385) for the fields in the list. I will also need to make sure that we've things aligned between the PatientService (DB) and the relevant FHIR Controller and Service classes.

**Comment by dixonwhitmire** [comment]
I'll take care of my [formatting issues](https://travis-ci.org/github/openemr/openemr/jobs/676037095) this morning.

**Comment by dixonwhitmire** [comment]
Hi @sjpadgett - Just checking in again . . .  I've finally resolved my formatting issues for the PSR-2 conventions. It looks like my remaining CI issues are unrelated to my changes, as the errors are due to the "E2E" suite failing (looks like Selenium tests).

In regards to next steps, should I continue to use this feature branch, or would you prefer that we use separate branches and merge each in turn? I just wanted to check and see what the next steps are for this bit of work, so that I can keep plugging along.

Oh for the unit tests, I can create those for the Patient Controller and Service components. It looks PHP Unit has support for "Doubles"/Mocking which we could utilize for the controller. I can use the "setup" and "teardown" hooks for the service component to add and remove sample patients from the database. It may be helpful to build out some tests for these API endpoints and DB services to ensure that things are working as expected.

Thanks!
Dixon

**Comment by sjpadgett** [comment]
> CI issues are unrelated to my changes

Yep, our Travis guru's are aware but won't deter us.


> In regards to next steps

I find when i'm in development mode that merging in parts is easier and saves against conflicts later. Because we're active with FHIR resources, i'd recommend bringing in in stages. Tests is a good example of a new PR.


> For the unit tests,

I'm afraid i'm not up on writing tests but, seems you are and do how you see fit. Perhaps @bradymiller has some thoughts as he has been pushing more tests development.

I'll review Travis then merge. This piece of FHIR is what i've been looking forward to more than anything and think it will be the most useful part of our FHIR offering( for our providers). Maybe even spur on some institutional development.
Thanks for the contribution.

**Comment by sjpadgett** [comment]
@bradymiller  hmm, did we recently update our composer dependencies versions?

**Comment by bradymiller** [comment]
hi @dixonwhitmire and @sjpadgett ,
All for any automated testing and the sky is the limit on this; currently have neat mechanisms in place that runs on any mysql/mariadb/php versions imaginable; also can be run from the easy dev docker which should make automated testing rather easy to add; would be awesome to have something that tested the api/fhir endpoints (note that the api automated testing that is now done is empty :) ).

btw, excised the test that was intermittently failing travis.

Was planning a composer dependency and npm/js dependency update soon.

File: src/RestControllers/PatientRestController.php [hunk 3]
```diff
@@ -45,19 +62,29 @@ public function put($pid, $data)
         return RestControllerHelper::responseHandler($serviceResult, array("pid" => $pid), 200);
     }
 
-    public function getOne()
```
Comment (prondubuisi) [inline src/RestControllers/PatientRestController.php:48]: Hello @dixonwhitmire I am late to the party, any reasons for removing the public access modifier? looking to learn

Comment (stephenwaite) [reply src/RestControllers/PatientRestController.php:48]: hi @prondubuisi , think if you don't specify it defaults to `public` so it was kind of redundant

Comment (bradymiller) [reply src/RestControllers/PatientRestController.php:48]: hi @prondubuisi , @stephenwaite , @dixonwhitmire 
this is a good point. psr2 specs rec to explicitly state either public/private/protected.
https://rules.sonarsource.com/php/tag/psr2/RSPEC-1784

Comment (stephenwaite) [reply src/RestControllers/PatientRestController.php:48]: can't blame travis for this one :)
this could go
https://github.com/openemr/openemr/blob/master/ci/phpcs.xml#L9
or just move to psr 12 :)
https://www.php-fig.org/psr/psr-2/

Comment (prondubuisi) [reply src/RestControllers/PatientRestController.php:48]: Lol @stephenwaite even Psr-12 requires that visibility MUST be declared https://www.php-fig.org/psr/psr-12/#44-methods-and-functions, that also helps for general consistency. Would have to remove the rule for PHPCS, thanks for pointing that out.

Comment (stephenwaite) [reply src/RestControllers/PatientRestController.php:48]: hi @prondubuisi, maybe we should we bump our `PSR-2` to `PSR-12`? We will need to checkout the other pieces of `ci/phpcs.xml`.

Comment (prondubuisi) [reply src/RestControllers/PatientRestController.php:48]: okay @stephenwaite. I am always in for new and shiny things!

Comment (sjpadgett) [reply src/RestControllers/PatientRestController.php:48]: > this is a good point. psr2 specs rec to explicitly state either public/private/protected.

I think this was just an oversight. I missed it too. I'll have @silvistershebin do a quick fix in his PR for procedures. Thanks for catching but my goodness, you'd think the world was ending:)

Comment (prondubuisi) [reply src/RestControllers/PatientRestController.php:48]: The old man with a walking stick always cracks me up  with his Sarcasm!

Comment (sjpadgett) [reply src/RestControllers/PatientRestController.php:48]: lol, yep and it has gotten me in trouble in the past. So, wouldn't rec for everyone.
I do love the smiley face though, it begs the audience to question umm "Was he joking" or did he really mean it.:) See.

Comment (bradymiller) [reply src/RestControllers/PatientRestController.php:48]: The world is ending!

tenor (10)

After the world ends, be good to place public back on these functions. This is something I usually force for classes that get placed in src/ directory (ie. PSR4 classes), but I was distracted with the end of the world coming and all.

Comment (dixonwhitmire) [reply src/RestControllers/PatientRestController.php:48]: Thanks for the discussion and lively feedback! 

On my next round of changes I will ensure that we have the appropriate access modifier on the components being modified. I will also go ahead and add `public` back in if it's not done by the time I submit a PR.

File: src/Services/PatientService.php (hunk 2 above)
Comment (yashrajbothra) [reply src/Services/PatientService.php:295]: image

Will this be covered in your next PR or I can push a quick fix :)

Comment (dixonwhitmire) [reply src/Services/PatientService.php:295]: @yashrajbothra - feel free to go ahead and push your fix!
